docs/fsharp/whats-new/fsharp-9.md

```diff
@@ -0,0 +1,432 @@
+```
+
+This makes it easier to use the F# immutable collections from C#, for example when you need their structural equality, which `System.Collections.Immutable` collections don't have.
+
+## Quality of life improvements
+
+### Parser recovery
+
+There have been continuous improvements in parser recovery, meaning that tooling (e.g. syntax highlighting) still works with code when you're in the middle of editing it and it may not be syntactically correct at all times.
+
+For example, the parser will now recover on unfinished `as` patterns, object expressions, enum case declarations, record declarations, complex primary constructor patterns, unresolved long identifiers, empty match clauses, missing union case fields, and missing union case field types.
+
+### Diagnostics
+
+Diagnostics, or understanding what the compiler doesn't like about your code, are an important part of the user experience with F#. There are a number of new or improved diagnostic messages or more precise diagnostic locations in F# 9.
+
+These include:
+- Ambiguous override method in object expression
+- Abstract members when used in non-abstract classes
+- When a property has the same name as a discriminated union case
+- Active pattern argument count mismatch
+- Unions with duplicated fields
+- Using `use!` with `and!` in computation expressions
+
+There is also a new compile-time error for classes with over 65,520 methods in generated [IL](https://learn.microsoft.com/dotnet/standard/managed-code#intermediate-language--execution). Such classes won't be loadable by the CLR and would result in a runtime error. (Not that you would ever really want to have so many methods, but there have been cases with generated code.)
+
+### Real visibility
+
+There is a quirk with how F# generates assemblies that results in private members being written to [IL](https://learn.microsoft.com/dotnet/standard/managed-code#intermediate-language--execution) as internal. This allows inappropriate access to private members from non-F# projects that have access to an F# project via [`InternalsVisibleTo`](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute).
+
+Now, there is an opt-in fix for this behavior available via the `--realsig+` compiler flag. Try it in your solution to see if any of your projects depend on this behavior. You can add it to your `.fsproj` files like this:
+
+<pre><code class="language-xml">&lt;PropertyGroup&gt;
+    &lt;OtherFlags&gt;--realsig+&lt;/OtherFlags&gt;
+&lt;/PropertyGroup&gt;
+```
+
+## Performance improvements
+
+### Optimized equality checks
+
+Equality checks are now faster and allocate less memory.
+
+For example:
+
+```fsharp
+[&lt;Struct&gt;]
+type MyId =
+    val Id: int
+    new id = { Id = id }
+
+let ids = Array.init 1000 MyId
+let missingId = MyId -1
+
+// used to box 1000 times, doesn't box anymore
+let _ = ids |> Array.contains missingId```
```

Comment:
People love BDN tables, I think this one could use a small copy paste of the results to really show how good the improvement is.